### PR TITLE
New mapping approach

### PR DESCRIPTION
**Pull request by nagyesta**
- Refreshes license information in file headers
- Rewrites API models to eliminate most of the custom mapping code
- Redefines API configuration options
- Simplifies fetching and mapping logic
- Updates dependency versions
- Reduces Java source level to 8 everywhere to eliminate JavaDoc warnings related to generated code
- Moves some in-line JSONs to class path resources to make tests more clean
- Adds assumptions to skip integration tests if API key is not set
- Adds assumptions to skip One Call API tests unless RUN_ONE_CALL env var is set
- Solves issue around UnsupportedOperationExceptions in certain terminators
- Adds option to use secure channel for picture URLs
- Adds new tests

Signed-off-by: Esta Nagy <nagyesta@gmail.com>

### CODE REVIEW

**Comment by nagyesta** [comment]
Hi @Prominence ,
Finally managed to complete the fixes and rebased. I think the code is ready for review now. I am not that familiar with Cirrus CI, but it appear to me, that there was some kind of an issue after the Gradle build completed. Could you please advise how we should fix it?
Thank you in advance!

File: src/main/java/com/github/prominence/openweathermap/api/model/generic/temperature/TemperatureValue.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        return value.setScale(DECIMAL_PLACES, RoundingMode.HALF_EVEN);
+    }
+
+    @JsonIgnore
+    public BigDecimal asCelsius() {
+        return value.subtract(KELVIN_TO_CELSIUS_OFFSET)
+                .setScale(DECIMAL_PLACES, RoundingMode.HALF_EVEN);
+    }
+
+    @JsonIgnore
```
Comment (nagyesta) [inline src/main/java/com/github/prominence/openweathermap/api/model/generic/temperature/TemperatureValue.java:57]: Missed test for this.

File: src/main/java/com/github/prominence/openweathermap/api/model/roadrisk/RoadRiskModel.java [hunk 2]
```diff
@@ -20,27 +20,35 @@
...
-        JsonNode weatherNode = p.getCodec().readTree(p);
-        final WeatherState weatherState = new WeatherState(
-                weatherNode.get("id").asInt(),
-                weatherNode.get("main").asText(),
-                weatherNode.get("description").asText()
-        );
-        weatherState.setIconId(weatherNode.get("icon").asText());
-        return weatherState;
+    public Coordinates getCoordinates() {
+        return new Coordinates(coord.get(0), coord.get(1));
```
Comment (nagyesta) [inline src/main/java/com/github/prominence/openweathermap/api/model/roadrisk/RoadRiskModel.java:52]: Missed test for this.

File: src/main/java/com/github/prominence/openweathermap/api/model/generic/temperature/TemperatureValue.java (hunk 1 above)
Comment (nagyesta) [reply src/main/java/com/github/prominence/openweathermap/api/model/generic/temperature/TemperatureValue.java:57]: Fixed.

File: src/main/java/com/github/prominence/openweathermap/api/model/roadrisk/RoadRiskModel.java (hunk 2 above)
Comment (nagyesta) [reply src/main/java/com/github/prominence/openweathermap/api/model/roadrisk/RoadRiskModel.java]: Fixed.

**Comment by nagyesta** [comment]
> Hi @Prominence , Finally managed to complete the fixes and rebased. I think the code is ready for review now. I am not that familiar with Cirrus CI, but it appear to me, that there was some kind of an issue after the Gradle build completed. Could you please advise how we should fix it? Thank you in advance!

Never mind, it was most probably only a temporary glitch.

**Review by Prominence — CHANGES_REQUESTED** [review]
Please, review my comment in the previous PR. I've missed your answer, sorry.

File: src/main/java/com/github/prominence/openweathermap/api/core/net/HttpURLConnectionBasedHttpClient.java [hunk 3]
```diff
@@ -108,19 +140,12 @@ public String executePostRequest(String url, String body) {
...
-    private static String convertInputStreamToString(InputStream inputStream) {
-        StringBuilder result = new StringBuilder();
-
+    private String convertInputStreamToString(InputStream inputStream) {
         try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8))) {
-            String line;
-            while ((line = reader.readLine()) != null) {
-                result.append(line);
-            }
+            return reader.lines().collect(Collectors.joining("\n"));
```
[the remark this answers is not in the document]
Comment (nagyesta) [reply src/main/java/com/github/prominence/openweathermap/api/core/net/HttpURLConnectionBasedHttpClient.java:145]: Thanks, I will fix this ASAP!

**Comment by nagyesta** [comment]
> Please, review my comment in the previous PR. I've missed your answer, sorry.

No worries! I think this must be the only open question from that PR: https://github.com/Prominence/openweathermap-java-api/pull/40#discussion_r984060074
If you are fine with option 1, I can easily add it to this PR when I will fix the comment above. Alternatively, this decision can be left for a future update as well, if you are not ready to decide yet.
Either way, please let me know what to do!

File: src/main/java/com/github/prominence/openweathermap/api/core/net/HttpURLConnectionBasedHttpClient.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (nagyesta) [reply src/main/java/com/github/prominence/openweathermap/api/core/net/HttpURLConnectionBasedHttpClient.java]: Done.

**Comment by nagyesta** [comment]
> > Please, review my comment in the previous PR. I've missed your answer, sorry.
> 
> No worries! I think this must be the only open question from that PR: [#40 (comment)](https://github.com/Prominence/openweathermap-java-api/pull/40#discussion_r984060074) If you are fine with option 1, I can easily add it to this PR when I will fix the comment above. Alternatively, this decision can be left for a future update as well, if you are not ready to decide yet. Either way, please let me know what to do!

Based on the discussion on the other PR, I have updated the code to use option 1: `.asHTML(UnitSystem)`/ `.asJSON(UnitSystem)`/ `.asXML(UnitSystem)`. Created a shorthand as well to avoid forcing developers to always populate default values as well, so there are parameterless pairs of these methods which are only calling the longer form with `STANDARD`.

**Review by Prominence — APPROVED** [review]
LGTM.

Can I merge it now?

**Comment by nagyesta** [comment]
Thank you for the review! Yes please, feel free to merge!

**Comment by Prominence** [comment]
> Thank you for the review! Yes please, feel free to merge!

Do you want me to release the new version with your changes ASAP, or can I implement other API types and then release?

P.S. Big thanks for your work!

**Comment by nagyesta** [comment]
Thank you, I am not in a hurry, please handle the release as you wish!

**Pull request merged**